runtime/cudaq/dynamics/callback.h

```diff
@@ -0,0 +1,112 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2022 - 2025 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#pragma once
+
+#include "cudaq/qis/state.h"
+#include "cudaq/utils/tensor.h"
+
+#include <complex>
+#include <functional>
+#include <iostream>
+#include <unordered_map>
+#include <string>
+#include <vector>
+
+namespace cudaq {
+
+class ScalarCallbackFunction {
+private:
+  // The user provided callback function that takes a map of complex 
+  // parameters.
+  std::function<std::complex<double>(const std::unordered_map<std::string, std::complex<double>>&)> _callback_func;
+
+public:
+  template <typename Callable>
+  ScalarCallbackFunction(Callable &&callable) {
+    static_assert(
+        std::is_invocable_r_v<std::complex<double>, Callable,
+                              const std::unordered_map<std::string, std::complex<double>>&>,
+        "Invalid callback function. Must have signature std::complex<double>("
+        "const std::unordered_map<std::string, std::complex<double>>&)");
+    _callback_func = std::forward<Callable>(callable);
+  }
+
+  // copy constructor
+  ScalarCallbackFunction(const ScalarCallbackFunction &other);
+
+  // move constructor.
+  ScalarCallbackFunction(ScalarCallbackFunction &&other);
+
+  // assignment operator
+  ScalarCallbackFunction& operator=(const ScalarCallbackFunction &other);
+
+  // move assignment operator
+  ScalarCallbackFunction& operator=(ScalarCallbackFunction &&other);
```

Comment:
The implementations in callback.cpp make me think they can be replaced with the default implementations that the compiler can provide (i.e. no special logic in there).
 ```suggestion
  // copy constructor
  ScalarCallbackFunction(const ScalarCallbackFunction &other) = default;

  // move constructor.
  ScalarCallbackFunction(ScalarCallbackFunction &&other) = default;

  // assignment operator
  ScalarCallbackFunction& operator=(const ScalarCallbackFunction &other) = default;

  // move assignment operator
  ScalarCallbackFunction& operator=(ScalarCallbackFunction &&other) = default;
```